### PR TITLE
Pin dart-lang/native dependencies

### LINKED ISSUE #137418: flutter_module_test frequent failures in build.dart
This appears to be a bug in the packages depended on by the tests ffi `build.dart` generated by https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/package_ffi/build.dart.tmpl#L1

@dcharkes 

```
[2023-10-27 02:46:19.229580] [STDOUT] stdout:                        build.dart returned with exit code: 255.
[2023-10-27 02:46:19.229586] [STDOUT] stdout:                        To reproduce run:
[2023-10-27 02:46:19.229593] [STDOUT] stdout:                        /opt/s/w/ir/x/w/recipe_cleanup/tmp8mlkod7v/flutter sdk/bin/dart --packages=/opt/s/w/ir/x/t/flutter_module_test.NhBYM6/hello/.dart_tool/package_config.json /opt/s/w/ir/x/t/flutter_module_test.NhBYM6/ffi_package/build.dart --config=/opt/s/w/ir/x/t/flutter_module_test.NhBYM6/hello/.dart_tool/native_assets_builder/e5f052c3025f3a319f6dd3717bed6fa6/config.yaml
[2023-10-27 02:46:19.229599] [STDOUT] stdout:                        stderr:
[2023-10-27 02:46:19.229605] [STDOUT] stdout:                        Unhandled exception:
[2023-10-27 02:46:19.229611] [STDOUT] stdout:                        Unsupported operation: Cannot add to an unmodifiable list
[2023-10-27 02:46:19.229617] [STDOUT] stdout:                        #0      UnmodifiableListMixin.addAll (dart:_internal/list.dart:129:5)
[2023-10-27 02:46:19.229623] [STDOUT] stdout:                        #1      CBuilder.run (package:native_toolchain_c/src/cbuilder/cbuilder.dart:276:45)
[2023-10-27 02:46:19.229629] [STDOUT] stdout:                        <asynchronous suspension>
[2023-10-27 02:46:19.229635] [STDOUT] stdout:                        #2      main (file:///opt/s/w/ir/x/t/flutter_module_test.NhBYM6/ffi_package/build.dart:18:3)
[2023-10-27 02:46:19.229641] [STDOUT] stdout:                        <asynchronous suspension>
```

https://github.com/dart-lang/native/blob/main/pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart#L276

### PR DESCRIPTION
**Pull request by dcharkes**
Pin the dependencies from dart-lang/native to a specific version during testing (rather than having them auto-upgrade during pub resolution). This will prevent tests using the template to start failing if a bad version is published to pub.

Closes: https://github.com/flutter/flutter/issues/137418

Also bumps dep in flutter_tools.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by christopherfujino — APPROVED** [review]
This LGTM.

> Maybe we should separate template and test? (Or use the template in the test but pin the deps before running pub get.)

Either of these would also make sense. It's just if we pin the deps while testing before pub get, we'd probably want to make sure that the autoroller is keeping the testing pins updated.

**Comment by dcharkes** [comment]
> This LGTM.
> 
> > Maybe we should separate template and test? (Or use the template in the test but pin the deps before running pub get.)
> 
> Either of these would also make sense. It's just if we pin the deps while testing before pub get, we'd probably want to make sure that the autoroller is keeping the testing pins updated.

The auto roller is not updating pubspec.yaml.tmpl files at all at the moment I believe. I've been manually submitting PRs to bump package:ffi and package:ffigen every once in a while. Do we track that work already somewhere?

**Comment by christopherfujino** [comment]
> > This LGTM.
> > > Maybe we should separate template and test? (Or use the template in the test but pin the deps before running pub get.)
> > 
> > 
> > Either of these would also make sense. It's just if we pin the deps while testing before pub get, we'd probably want to make sure that the autoroller is keeping the testing pins updated.
> 
> The auto roller is not updating pubspec.yaml.tmpl files at all at the moment I believe. I've been manually submitting PRs to bump package:ffi and package:ffigen every once in a while. Do we track that work already somewhere?

Just filed: https://github.com/flutter/flutter/issues/137643

**Comment by dcharkes** [comment]
> Or use the template in the test but pin the deps before running pub get.

Updated the PR to do this.

(Having the template pin the deps will lead to issues later if there are multiple packages that have used the template but have other minor versions in their dependencies. E.g. the packages are packages meant to be imported by others, not just apps that are only ever a root package.)

File: packages/flutter_tools/test/integration.shard/native_assets_test.dart [hunk 1]
```diff
@@ -394,9 +396,35 @@ Future<Directory> createTestProject(String packageName, Directory tempDirectory)
...
+      'pub',
+      'get',
+    ],
+    workingDirectory: packageDirectory.path,
+  );
+  if (result2.exitCode != 0) {
+    throw Exception(
+      'flutter create failed: ${result2.exitCode}\n${result2.stderr}\n${result2.stdout}',
+    );
+  }
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/native_assets_test.dart:415]: we did this so often we created a custom matcher for it:

```suggestion
  expect(result2, const ProcessResultMatcher());
```

File: packages/flutter_tools/test/integration.shard/native_assets_test.dart [hunk 2]
```diff
@@ -394,9 +396,35 @@ Future<Directory> createTestProject(String packageName, Directory tempDirectory)
...
+
   return packageDirectory;
 }
 
+Future<void> pinDependencies(File pubspecFile) async {
+  expect(pubspecFile, exists);
+  final String oldPubspec = await pubspecFile.readAsString();
+  final String newPubspec = oldPubspec.replaceAll(': ^', ': ');
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/native_assets_test.dart:423]: call me paranoid, but:

```suggestion
  final String newPubspec = oldPubspec.replaceAll(RegExp(r':\s*\^'), ': ');
```

**Pull request merged**